synapse/lib/modelrev.py

```diff
@@ -245,6 +248,496 @@ async def save():
+                if self.typeHasNdefs(typedef.tcache[name]):
+                    return True
+
+        elif isinstance(typedef, s_types.Array):
+            return self.typeHasNdefs(typedef.arraytype)
+
+        elif isinstance(typedef, (s_types.Ndef, s_types.Edge, s_types.NodeProp)):
+            return True
+
+        return False
+
+    def getElementsByStortype(self, stortype):
+        '''
+        Get the names of form, prop and tagprop model elements which use a specific
+        stortype, or refer to other nodes via ndefs/nodeprops.
+        '''
+        forms = set()
+        props = set()
+        tagprops = set()
+
+        for name, prop in self.core.model.props.items():
+            if not prop.isform and prop.univ:
+                continue
+
+            if self.typeHasStortype(prop.type, stortype) or self.typeHasNdefs(prop.type):
+                if prop.isform:
+                    forms.add(name)
+                else:
+                    props.add(name)
+
+        for tname, tprop in self.core.model.tagprops.items():
+            if self.typeHasStortype(tprop.type, stortype) or self.typeHasNdefs(tprop.type):
+                tagprops.add(tname)
+
+        return (forms, props, tagprops)
+
+    async def updateProps(self, props, layers, skip=None):
+        '''
+        Lift and re-norm prop values for the specified props in the specified layers.
+        '''
+        hugestorv1 = s_layer.StorTypeHugeNumV1(layers[0], s_layer.STOR_TYPE_HUGENUM)
+
+        for prop in props:
+            ptyp = self.core.model.props[prop]
+            form = ptyp.form
+
+            if form:
+                form = form.name
+            if skip is not None and (form and form in skip):
+                continue
+
+            pname = ptyp.name
+            stortype = ptyp.type.stortype
+            isarray = False
+            if stortype & s_layer.STOR_FLAG_ARRAY:
+                isarray = True
+                realtype = stortype & 0x7fff
+                if realtype == s_layer.STOR_TYPE_HUGENUM:
+                    realstor = hugestorv1
+                else:
+                    realstor = layers[0].stortypes[realtype]
+
+            def arrayindx(valu):
+                return set([realstor.indx(aval)[0] for aval in valu])
+
+            for layr in layers:
+
+                delindx = []
+                delarrayindx = []
+                nodeedits = []
+                meta = {'time': s_common.now(), 'user': self.core.auth.rootuser.iden,
+                        'migr': {'delpropindx': delindx, 'delproparrayindx': delarrayindx}}
+
+                async def save():
+                    await layr.storNodeEdits(nodeedits, meta)
+                    nodeedits.clear()
+                    delindx.clear()
+                    delarrayindx.clear()
+
+                try:
+                    indxby = s_layer.IndxByProp(layr, form, pname)
+                except s_exc.NoSuchAbrv:
+                    continue
```

Comment:
please add a one-liner comment explaining the context. I suspect it's just "no instances of that prop exist in the layer" but would be nice to have it for future self